include/rcutils/error_handling.h

```diff
@@ -22,58 +22,123 @@ extern "C"
 {
 #endif
 
+#ifndef __STDC_WANT_LIB_EXT1__
+#define __STDC_WANT_LIB_EXT1__ 1  // indicate we would like strnlen_s if available
+#endif
+#include <assert.h>
 #include <stdbool.h>
 #include <stddef.h>
+#include <stdint.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 
 #include "rcutils/allocator.h"
-#include "rcutils/format_string.h"
 #include "rcutils/macros.h"
+#include "rcutils/snprintf.h"
 #include "rcutils/types/rcutils_ret.h"
 #include "rcutils/visibility_control.h"
 
+#ifdef __STDC_LIB_EXT1__
+// Limit the buffer size in the `fwrite` call to give an upper bound to buffer overrun in the case
+// of non-null terminated `msg`.
+#define RCUTILS_SAFE_FWRITE_TO_STDERR(msg) \
+  do {fwrite(msg, sizeof(char), strnlen_s(msg, 4096), stderr);} while (0)
+#else
+#define RCUTILS_SAFE_FWRITE_TO_STDERR(msg) \
+  do {fwrite(msg, sizeof(char), strlen(msg), stderr);} while (0)
+#endif
+
+// fixed constraints
+#define RCUTILS_ERROR_STATE_LINE_NUMBER_STR_MAX_LENGTH 20  // "18446744073709551615"
+#define RCUTILS_ERROR_FORMATTING_CHARACTERS 6  // ', at ' + ':'
+
+// max formatted string length
+#define RCUTILS_ERROR_MESSAGE_MAX_LENGTH 1024
+
+// adjustable max length for user defined error message
+// remember "chained" errors will include previously specified file paths
+// e.g. "some error, at /path/to/a.c:42, at /path/to/b.c:42"
+#define RCUTILS_ERROR_STATE_MESSAGE_MAX_LENGTH 768
+// with RCUTILS_ERROR_STATE_MESSAGE_MAX_LENGTH = 768, RCUTILS_ERROR_STATE_FILE_MAX_LENGTH == 228
+#define RCUTILS_ERROR_STATE_FILE_MAX_LENGTH \
+  RCUTILS_ERROR_MESSAGE_MAX_LENGTH - \
+  RCUTILS_ERROR_STATE_MESSAGE_MAX_LENGTH - \
+  RCUTILS_ERROR_STATE_LINE_NUMBER_STR_MAX_LENGTH - \
+  RCUTILS_ERROR_FORMATTING_CHARACTERS - \
+  1
```

Comment:
nit Recommend parenthesis to avoid order of operation issues like `2 * RCUTILS_ERROR_STATE_FILE_MAX_LENGTH`